### PR TITLE
adding feedback to update event of collection

### LINKED ISSUE #3700: Backbone update event on collection should return the newly added or removed model(s)

### PR DESCRIPTION
**Pull request by elgubenis**
collection's update event receives 2 additional attributes, in case of removal of existing models, previousModels is an array containing the removed models which where referenced to the collection before .set was called. The second attribute, addedModels contains an array of models which were added to the collection calling the .set method.

Fixes #3700

### CODE REVIEW

File: backbone.js [hunk 1]
```diff
@@ -890,15 +890,19 @@
...
+      // Unless silenced, it's time to fire all appropriate add/update/sort events.
       if (!options.silent) {
         var addOpts = at != null ? _.clone(options) : options;
         for (var i = 0; i < toAdd.length; i++) {
           if (at != null) addOpts.index = at + i;
           (model = toAdd[i]).trigger('add', model, this, addOpts);
         }
+        if (toAdd.length || previousModels.length) {
+          if (toAdd.length) options.addedModels = toAdd;
+          if (previousModels.length) options.previousModels = previousModels;
```
Comment (megawac) [inline backbone.js]: is `previousModels` the right term or should it be removed models?

File: backbone.js [hunk 2]
```diff
@@ -867,7 +867,7 @@
         for (var i = 0; i < this.length; i++) {
           if (!modelMap[(model = this.models[i]).cid]) toRemove.push(model);
         }
-        if (toRemove.length) this._removeModels(toRemove, options);
+        if (toRemove.length) previousModels = this._removeModels(toRemove, options);
```
Comment (jridgewell) [inline backbone.js]: Everything in `toRemove` has already been [verified](https://github.com/jashkenas/backbone/pull/3711/files#diff-0d56d0d310de7ff18b3cef9c2f8f75dcR868) from the current models, shouldn't need to grab `#_removeModels` return.

**Comment by megawac** [comment]
:+1:

**Comment by elgubenis** [comment]
i would also like to include mergedModels but i think that wont be possible (see #3708)

File: backbone.js (hunk 1 above)
Comment (elgubenis) [inline backbone.js]: i thought it should be removedModels, but i saw that .reset returns previousModels aswel, so i think keeping both the same would not be wrong.

Comment (jridgewell) [inline backbone.js]: Mind calling these `options.add` and `options.remove`? It mirrors the `add` and `remove` event names.

File: backbone.js (hunk 2 above)
Comment (elgubenis) [inline backbone.js]: so should it previosModels equal to toRemove then?

File: backbone.js (hunk 1 above)
Comment (megawac) [inline backbone.js]: I think past tense works better here, `added` & `removed`?

**Comment by jridgewell** [comment]
Mind adding the removed models to the `update` event in `#remove`?

**Comment by elgubenis** [comment]
sure

File: backbone.js (hunk 2 above)
Comment (jridgewell) [inline backbone.js]: Yup.

File: backbone.js (hunk 1 above)
Comment (jridgewell) [inline backbone.js]: I'm pro past tense, but nothing else in Backbone's public API uses it. @jashkenas, @braddunbar, or @akre54 want to weigh in?

**Comment by jridgewell** [comment]
Looks good to me, the only thing I want feedback on is the `addedModels`/`removedModels` naming.

File: backbone.js (hunk 1 above)
Comment (elgubenis) [inline backbone.js]: options.add and options.remove are already set, because they are the options passed to the .set

**Comment by elgubenis** [comment]
yeah, i didnt like it either, but since .reset returns previousModels, i thought they should be named alike, well lets wait for the feedback.

**Comment by elgubenis** [comment]
should i write some tests to see that the events in .remove return the removed models which it should return? as well as the removed/previous and added ones for .set? or is that not necessary

**Comment by jridgewell** [comment]
> should i write some tests to see that the events in .remove return the removed models which it should return? as well as the removed/previous and added ones for .set?

:+1:

File: test/collection.js [hunk 3]
```diff
@@ -1671,4 +1671,98 @@
     collection.invoke('method', 1, 2, 3);
   });
 
+  test("#3711 - remove's `update` event returns one removed model", function() {
+    var model = new Backbone.Model({id: 1, title: 'First Post'});
+    var collection = new Backbone.Collection([model]);
+    collection.on('update', function(context, options) {
+      if (options.removedModels[0] !== model) {
+        ok(false);
```
Comment (akre54) [inline test/collection.js]: Try just `ok(options.removedModels[0] === model)` or even better, `strictEqual(options.removedModels[0], model)`

Comment (elgubenis) [inline test/collection.js]: like that? ```js !strictEqual(options.removedModels[0], model) ``` i guess?

Comment (megawac) [inline test/collection.js]: notStrictEqual()

File: backbone.js [hunk 4]
```diff
@@ -890,15 +891,19 @@
...
+      // Unless silenced, it's time to fire all appropriate add/update/sort events.
       if (!options.silent) {
         var addOpts = at != null ? _.clone(options) : options;
         for (var i = 0; i < toAdd.length; i++) {
           if (at != null) addOpts.index = at + i;
           (model = toAdd[i]).trigger('add', model, this, addOpts);
         }
+        if (toAdd.length || toRemove.length) {
+          if (toAdd.length) options.addedModels = toAdd;
+          if (toRemove.length) options.previousModels = toRemove;
```
Comment (megawac) [inline backbone.js]: why not just always set `addedModels` and `previousModels` on options and trigger updates only if theres a `length` change. Seems mostly harmless (actually better IMO as the user won't have to check if `toAdd` is an array or not)

Comment (elgubenis) [inline backbone.js]: i saw the same behavior in the set method, previousModels is only set when there are previousModels, so i just copied that, normally i would also be in favor of having it all the time, but i didnt want to change how the other methods do work.

Comment (jridgewell) [inline backbone.js]: The only other `previousModels` always attaches. I think attaching an empty array here is fine as well.

Also, can we rename this `removedModels`? `previousModels` makes sense in `#reset` where every model is "previous", but here only a subset are being removed.

Comment (elgubenis) [inline backbone.js]: ok

**Comment by elgubenis** [comment]
sorry, wrote the test a little too late

**Comment by jridgewell** [comment]
LGTM.

**Comment by jashkenas** [comment]
Naming isn't quite there yet — but more importantly without some sort of `mergedModels`, this change feels incomplete. Having the full bookkeeping of changes to models isn't possible?

**Comment by jridgewell** [comment]
We can add a `toMerge` [here](https://github.com/jashkenas/backbone/blob/master/backbone.js#L833-L834), and push to it [here](https://github.com/jashkenas/backbone/blob/master/backbone.js#L855-L860).

**Comment by jashkenas** [comment]
Alright. And there isn't a big perf cost from the extra bookkeeping?

Naming wise `added`, `removed`, `merged` sounds good. Perhaps even under a sub-object if necessary.

**Comment by jridgewell** [comment]
> Alright. And there isn't a big perf cost from the extra bookkeeping?

With everything else this method is doing, I don't image one.

**Comment by elgubenis** [comment]
hi, im working on the PR, how could the sub-object be named? changes, status, updates, move(s)/ments, details?

**Comment by elgubenis** [comment]
i will dub it 'changes' for now if thats ok

**Comment by jridgewell** [comment]
@elgubenis: You'll have to rebase off of the current master for this to merge cleanly.

**Comment by elgubenis** [comment]
will write some tests for the merging

**Comment by jridgewell** [comment]
LGTM.

**Comment by jridgewell** [comment]
@elgubenis: Mind rebasing? I'll merge this afterwards.

**Comment by Mordentware** [comment]
Just wanted to add that I was after this exact functionality, cheers guys for implementing. Any chance of getting the conflicts sorted out?

**Comment by elgubenis** [comment]
hi, will do.

File: backbone.js [hunk 5]
```diff
@@ -912,7 +917,13 @@
           model.trigger('add', model, this, options);
         }
         if (sort || orderChanged) this.trigger('sort', this, options);
-        if (toAdd.length || toRemove.length) this.trigger('update', this, options);
+        if (toAdd.length || toRemove.length || toMerge.length) {
+          options.changes = {};
```
Comment (akre54) [inline backbone.js]: minor style nit:

```js
options.changes = {
  added: toAdd,
  removed: toRemove,
  merged: toMerge
}
```

File: backbone.js [hunk 6]
```diff
@@ -808,7 +808,10 @@
       var singular = !_.isArray(models);
       models = singular ? [models] : _.clone(models);
       var removed = this._removeModels(models, options);
-      if (!options.silent && removed) this.trigger('update', this, options);
+      if (!options.silent && removed) {
+        options.removedModels = removed;
```
Comment (akre54) [inline backbone.js]: Why is this one `options.removedModels` but below it's `options.changes.removed`?

**Comment by megawac** [comment]
@elgubenis mind addressing @akre54 feedback and we'll merge? Thanks

**Pull request closed without merging**